pow/validator/src/main/java/org/ethereum/beacon/pow/validator/ValidatorRegistrationServiceImpl.java

```diff
@@ -0,0 +1,298 @@
+import java.util.concurrent.Executors;
+import java.util.function.Function;
+
+import static org.ethereum.beacon.core.spec.SignatureDomains.DEPOSIT;
+
+public class ValidatorRegistrationServiceImpl implements ValidatorRegistrationService {
+  private final TransactionBuilder transactionBuilder;
+  private final TransactionGateway transactionGateway;
+  private final DepositContract depositContract;
+  private final Publisher<ObservableBeaconState> observablePublisher;
+  private final SingleValueSource<RegistrationStage> stagePersistence;
+
+  private final SpecHelpers specHelpers;
+  private final ObjectHasher<Hash32> sszHasher;
+  private final Serializer sszSerializer;
+
+  private Disposable depositSubscription = null;
+  private ValidatorService validatorService = null;
+  private RegistrationStage currentStage = null;
+  private ValidatorRecord validatorRecord = null;
+
+  // Validator
+  private MessageSigner<BLSSignature> signer;
+  private BLSPubkey pubKey;
+  private Hash32 withdrawalCredentials = null;
+  private Gwei amount = null;
+  private Address eth1From = null;
+  private BytesValue eth1PrivKey = null;
+
+  private ExecutorService executor;
+
+  public ValidatorRegistrationServiceImpl(
+      TransactionBuilder transactionBuilder,
+      TransactionGateway transactionGateway,
+      DepositContract depositContract,
+      Publisher<ObservableBeaconState> observablePublisher,
+      SingleValueSource<RegistrationStage> registrationStagePersistence,
+      SpecHelpers specHelpers) {
+    this.transactionBuilder = transactionBuilder;
+    this.transactionGateway = transactionGateway;
+    this.depositContract = depositContract;
+    this.observablePublisher = observablePublisher;
+    this.stagePersistence = registrationStagePersistence;
+    this.specHelpers = specHelpers;
+    Function<BytesValue, Hash32> hashFunction = Hashes::keccak256;
+    sszHasher = SSZObjectHasher.create(hashFunction);
+    sszSerializer = Serializer.annotationSerializer();
+  }
+
+  @Override
+  public void start(
+      MessageSigner<BLSSignature> signer,
+      BLSPubkey pubKey,
+      @Nullable Hash32 withdrawalCredentials,
+      @Nullable Gwei amount,
+      @Nullable Address eth1From,
+      @Nullable BytesValue eth1PrivKey) {
+    this.signer = signer;
+    this.pubKey = pubKey;
+    this.executor =
+        Executors.newSingleThreadExecutor(
+            runnable -> {
+              Thread t = new Thread(runnable, "validator-registration-service");
+              t.setDaemon(true);
+              return t;
+            });
+    actionCircle();
+  }
+
+  private void actionCircle() {
+    RegistrationStage currentStage = getCurrentStage();
+    getCurrentStageFunction(currentStage).run();
+  }
+
+  private void changeCurrentStage(RegistrationStage newStage) {
+    this.currentStage = newStage;
+    stagePersistence.set(currentStage);
+    actionCircle();
+  }
+
+  private RegistrationStage getCurrentStage() {
```

Comment:
This method doesn't do a check if this validator has already been registered on contract. It might be a useful thing.

Reply:
How about the approach I made?